### PR TITLE
[TS Migration] Migrate withWindowDimensions and Modal directories to TypeScript

### PR DESCRIPTION
**Pull request by JKobrynski**
### Fixed Issues

$ #25139 
PROPOSAL: N/A

### Tests

1. Log into the app
2. Test different Modals and make sure they work/look correctly. Some examples of where you can find Modals:
- Attachment modal (when you want to send an attachment in a conversation)
- Delete comment modal

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as **Tests** section above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android.webm](https://github.com/Expensify/App/assets/44364829/a36371d4-96e9-4d7f-adf4-94f9158bfa7a)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[chrome.webm](https://github.com/Expensify/App/assets/44364829/50744f00-69b7-480c-9448-7722ea2c18e4)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/44364829/1eefdecf-799e-4d69-a46e-e915cb7e7913

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/44364829/37d157b3-77c0-4e5c-a309-4263e74a7fe4

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/44364829/6471fd87-47e9-4c4e-ba6f-bb90067d8719

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/44364829/9f253834-f627-4e08-85c7-9d85c8ed5004

</details>

### CODE REVIEW

File: src/components/Modal/BaseModal.tsx [hunk 1]
```diff
@@ -110,13 +98,10 @@ function BaseModal({
         if (shouldSetModalVisibility) {
             Modal.setModalVisibility(true);
         }
-        onModalShow();
+        onModalShow?.();
     };
 
-    const handleBackdropPress = (e) => {
-        if (e && e.key === CONST.KEYBOARD_SHORTCUTS.ENTER.shortcutKey) {
```
Comment (kubabutkiewicz) [inline src/components/Modal/BaseModal.tsx:117]: We we are removing that?

File: src/components/Modal/BaseModal.tsx [hunk 2]
```diff
@@ -1,57 +1,45 @@
...
-    statusBarTranslucent,
-    onLayout,
-    avoidKeyboard,
-    forwardedRef,
-    children,
-}) {
+import BaseModalProps from './types';
+
+function BaseModal(
+    {
```
Comment (barttom) [inline src/components/Modal/BaseModal.tsx:18]: is linter move this bracket to the next line? I think should be like `function BaseModal({`

File: src/components/Modal/BaseModal.tsx [hunk 3]
```diff
@@ -61,15 +49,15 @@ function BaseModal({
...
+     * @param callHideCallback - Should we call the onModalHide callback
      */
     const hideModal = useCallback(
         (callHideCallback = true) => {
             if (shouldSetModalVisibility) {
                 Modal.setModalVisibility(false);
             }
             if (callHideCallback) {
-                onModalHide();
+                onModalHide?.();
```
Comment (barttom) [inline src/components/Modal/BaseModal.tsx:60]: You defined default prop - in taht case conditional calling is needed?

File: src/components/Modal/BaseModal.tsx [hunk 4]
```diff
@@ -110,13 +98,10 @@ function BaseModal({
         if (shouldSetModalVisibility) {
             Modal.setModalVisibility(true);
         }
-        onModalShow();
+        onModalShow?.();
```
Comment (barttom) [inline src/components/Modal/BaseModal.tsx:101]: You defined default prop as an empty fucntion - in taht case conditional calling is needed?

File: src/components/Modal/BaseModal.tsx (hunk 1 above)
Comment (JKobrynski) [reply src/components/Modal/BaseModal.tsx:117]: It looks to me like this event handler doesn't accept any parameters

File: src/components/withWindowDimensions/index.tsx [hunk 5]
```diff
@@ -43,7 +40,7 @@ function WindowDimensionsProvider(props) {
     });
 
     useEffect(() => {
-        const onDimensionChange = (newDimensions) => {
+        const onDimensionChange = (newDimensions: {window: ScaledSize}) => {
```
Comment (barttom) [inline src/components/withWindowDimensions/index.tsx:43]: maybe worth to define newDimension type, looks like You used it twice

File: src/components/Modal/BaseModal.tsx (hunk 2 above)
Comment (JKobrynski) [reply src/components/Modal/BaseModal.tsx:18]: Yes it did, I think it's because of the ref (a second argument)

File: src/components/Modal/BaseModal.tsx (hunk 3 above)
Comment (JKobrynski) [reply src/components/Modal/BaseModal.tsx]: I don't think so, updated!

File: src/components/Modal/BaseModal.tsx (hunk 4 above)
Comment (JKobrynski) [reply src/components/Modal/BaseModal.tsx]: Updated

File: src/components/withWindowDimensions/index.tsx (hunk 5 above)
Comment (JKobrynski) [reply src/components/withWindowDimensions/index.tsx]: Great idea, done!

File: src/components/Modal/BaseModal.tsx (hunk 1 above)
Comment (fabioh8010) [reply src/components/Modal/BaseModal.tsx:117]: @JKobrynski Please test this extensively and confirm if the event is being sent or not (maybe in some platforms is being sent and others not) in order to avoid regressions

File: src/components/Modal/index.tsx [hunk 6]
```diff
@@ -25,12 +25,15 @@ function Modal(props) {
...
-        setStatusBarColor(isFullScreenModal ? themeColors.appBG : StyleUtils.getThemeBackgroundColor(statusBarColor));
-        props.onModalShow();
+
+        if (statusBarColor) {
+            const isFullScreenModal =
+                props.type === CONST.MODAL.MODAL_TYPE.CENTERED || props.type === CONST.MODAL.MODAL_TYPE.CENTERED_UNSWIPEABLE || props.type === CONST.MODAL.MODAL_TYPE.RIGHT_DOCKED;
+            setPreviousStatusBarColor(statusBarColor);
+            // If it is a full screen modal then match it with appBG, otherwise we use the backdrop color
+            setStatusBarColor(isFullScreenModal ? themeColors.appBG : StyleUtils.getThemeBackgroundColor(statusBarColor));
+            props.onModalShow();
```
Comment (fabioh8010) [inline src/components/Modal/index.tsx:35]: I think is quite dangerous to only call `onModalShow` if `statusBarColor` is defined, these things look unrelated

File: src/components/Modal/types.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        /** Callback method fired when the modal is shown */
+        onModalShow?: () => void;
+
+        /** Style of modal to display */
+        // type: PropTypes.oneOf(_.values(CONST.MODAL.MODAL_TYPE)),
+        type?: ValueOf<typeof CONST.MODAL.MODAL_TYPE>;
+
+        /** The anchor position of a popover modal. Has no effect on other modal types. */
+        popoverAnchorPosition?: {
```
Comment (fabioh8010) [inline src/components/Modal/types.ts:38]: Extract this to a separate type

File: src/components/withWindowDimensions/index.native.tsx [hunk 8]
```diff
@@ -76,30 +72,31 @@ function WindowDimensionsProvider(props) {
...
 
-WindowDimensionsProvider.propTypes = windowDimensionsProviderPropTypes;
 WindowDimensionsProvider.displayName = 'WindowDimensionsProvider';
 
 /**
- * @param {React.Component} WrappedComponent
- * @returns {React.Component}
+ * @param WrappedComponent
+ * @returns
  */
```
Comment (fabioh8010) [inline src/components/withWindowDimensions/index.native.tsx:80]: We can remove this comments

File: src/components/withWindowDimensions/index.tsx [hunk 9]
```diff
@@ -81,30 +78,31 @@ function WindowDimensionsProvider(props) {
...
 
-WindowDimensionsProvider.propTypes = windowDimensionsProviderPropTypes;
 WindowDimensionsProvider.displayName = 'WindowDimensionsProvider';
 
 /**
- * @param {React.Component} WrappedComponent
- * @returns {React.Component}
+ * @param WrappedComponent
+ * @returns
  */
```
Comment (fabioh8010) [inline src/components/withWindowDimensions/index.tsx:86]: We can remove this comments

File: src/components/Modal/BaseModal.tsx (hunk 1 above)
Comment (JKobrynski) [reply src/components/Modal/BaseModal.tsx:117]: Turns out event is actually present but only on web, I brought it back!

File: src/components/Modal/index.tsx (hunk 6 above)
Comment (JKobrynski) [reply src/components/Modal/index.tsx]: Right, I narrowed it down to wrap only what needs to be wrapped

File: src/components/Modal/types.ts (hunk 7 above)
Comment (JKobrynski) [reply src/components/Modal/types.ts:39]: Done!

File: src/components/withWindowDimensions/index.native.tsx (hunk 8 above)
Comment (JKobrynski) [reply src/components/withWindowDimensions/index.native.tsx]: Right, updated

File: src/components/withWindowDimensions/index.tsx (hunk 9 above)
Comment (JKobrynski) [reply src/components/withWindowDimensions/index.tsx]: Updated!

File: src/components/Modal/types.ts (hunk 7 above)
Comment (fabioh8010) [reply src/components/Modal/types.ts:39]: @JKobrynski I was talking about `popoverAnchorPosition` 🙂

Comment (JKobrynski) [reply src/components/Modal/types.ts]: Updated 😅

**Review by fabioh8010 — APPROVED** [review]
LGTM

File: src/components/withWindowDimensions/index.tsx [hunk 10]
```diff
@@ -81,30 +78,27 @@ function WindowDimensionsProvider(props) {
...
-                    // eslint-disable-next-line react/jsx-props-no-spreading
-                    {...windowDimensionsProps}
-                    // eslint-disable-next-line react/jsx-props-no-spreading
-                    {...props}
-                    ref={ref}
-                />
-            )}
-        </WindowDimensionsContext.Consumer>
-    ));
+export default function withWindowDimensions<TProps extends WindowDimensionsProps, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>) {
```
Comment (blazejkustra) [inline src/components/withWindowDimensions/index.tsx:83]: Add return type to this HOC

**Review by blazejkustra — APPROVED** [review]
One small comment, other than that LGTM!

File: src/components/withWindowDimensions/index.tsx (hunk 10 above)
Comment (JKobrynski) [reply src/components/withWindowDimensions/index.tsx]: Done!

**Comment by situchan** [comment]
I can help

**Comment by mountiny** [comment]
@situchan please go ahead

**Comment by situchan** [comment]
@blazejkustra sorry, planning to review in an hour

**Comment by situchan** [comment]
@blazejkustra can we fix [crash](https://github.com/Expensify/App/pull/28589#discussion_r1384767140) now? So I can complete review and jump into this one

**Comment by blazejkustra** [comment]
@BartoszGrajdek is on it, thank you!

**Comment by situchan** [comment]
Conflicts!

**Comment by JKobrynski** [comment]
@situchan conflicts resolved!

File: src/components/Modal/index.android.tsx [hunk 11]
```diff
@@ -15,19 +15,17 @@ AppState.addEventListener('blur', () => {
...
 // Only want to use useNativeDriver on Android. It has strange flashes issue on IOS
 // https://github.com/react-native-modal/react-native-modal#the-modal-flashes-in-a-weird-way-when-animating
-function Modal(props) {
+function Modal(props: BaseModalProps) {
     return (
         <BaseModal
-            useNativeDriver
             // eslint-disable-next-line react/jsx-props-no-spreading
             {...props}
+            useNativeDriver
```
Comment (situchan) [inline src/components/Modal/index.android.tsx:23]: Any reason for overwriting this prop? What if props has useNativeDriver=false?

**Comment by situchan** [comment]
We need Tests / QA Steps. And remoev `[No QA]` from title

**Comment by situchan** [comment]
Bug on delete modal:

bug

File: src/components/Modal/index.android.tsx (hunk 11 above)
Comment (JKobrynski) [reply src/components/Modal/index.android.tsx:23]: I must've added it by mistake, I removed it 😄

File: src/components/Modal/index.android.tsx [hunk 12]
```diff
@@ -15,10 +15,9 @@ AppState.addEventListener('blur', () => {
 
 // Only want to use useNativeDriver on Android. It has strange flashes issue on IOS
 // https://github.com/react-native-modal/react-native-modal#the-modal-flashes-in-a-weird-way-when-animating
-function Modal(props) {
+function Modal(props: BaseModalProps) {
     return (
         <BaseModal
-            useNativeDriver
```
Comment (situchan) [inline src/components/Modal/index.android.tsx:21]: This should be back. I just meant ordering

Comment (JKobrynski) [reply src/components/Modal/index.android.tsx:21]: Done

**Comment by situchan** [comment]
Please also check https://github.com/Expensify/App/pull/30338#issuecomment-1801607970

**Comment by JKobrynski** [comment]
@situchan prop brought back, bug fixed

File: src/components/Modal/index.tsx [hunk 13]
```diff
@@ -25,12 +25,17 @@ function Modal(props) {
...
-        setStatusBarColor(isFullScreenModal ? themeColors.appBG : StyleUtils.getThemeBackgroundColor(statusBarColor));
-        props.onModalShow();
+
+        if (statusBarColor) {
+            setPreviousStatusBarColor(statusBarColor);
+            // If it is a full screen modal then match it with appBG, otherwise we use the backdrop color
+            setStatusBarColor(isFullScreenModal ? themeColors.appBG : StyleUtils.getThemeBackgroundColor(statusBarColor));
+        }
+
+        props.onModalShow?.();
```
Comment (situchan) [inline src/components/Modal/index.tsx:38]: This used to work without `?.` What's the root cause?

File: src/components/Modal/index.tsx [hunk 14]
```diff
@@ -5,10 +5,10 @@ import * as StyleUtils from '@styles/StyleUtils';
 import themeColors from '@styles/themes/default';
 import CONST from '@src/CONST';
 import BaseModal from './BaseModal';
-import {defaultProps, propTypes} from './modalPropTypes';
+import BaseModalProps from './types';
 
-function Modal(props) {
-    const [previousStatusBarColor, setPreviousStatusBarColor] = useState();
+function Modal(props: BaseModalProps) {
```
Comment (situchan) [inline src/components/Modal/index.tsx:10]: Let's destructure props

File: src/components/Modal/index.ios.tsx [hunk 15]
```diff
@@ -1,9 +1,9 @@
 import React from 'react';
 import withWindowDimensions from '@components/withWindowDimensions';
 import BaseModal from './BaseModal';
-import {defaultProps, propTypes} from './modalPropTypes';
+import BaseModalProps from './types';
 
-function Modal(props) {
+function Modal(props: BaseModalProps) {
```
Comment (situchan) [inline src/components/Modal/index.ios.tsx:6]: Same here

File: src/components/Modal/index.tsx (hunk 14 above)
Comment (JKobrynski) [reply src/components/Modal/index.tsx]: Done!

File: src/components/Modal/index.ios.tsx (hunk 15 above)
Comment (JKobrynski) [reply src/components/Modal/index.ios.tsx]: Done!

**Comment by BartoszGrajdek** [comment]
@situchan kind bump 😄

**Comment by situchan** [comment]
@BartoszGrajdek please pull main. author checklist failing

**Comment by mountiny** [comment]
@situchan I would disregard the checklist for now, there as been recent changes to it

File: src/components/Modal/index.tsx [hunk 16]
```diff
@@ -5,13 +5,13 @@ import * as StyleUtils from '@styles/StyleUtils';
 import themeColors from '@styles/themes/default';
 import CONST from '@src/CONST';
 import BaseModal from './BaseModal';
-import {defaultProps, propTypes} from './modalPropTypes';
+import BaseModalProps from './types';
 
-function Modal(props) {
-    const [previousStatusBarColor, setPreviousStatusBarColor] = useState();
+function Modal({fullscreen, onModalHide, type, onModalShow, children, ...rest}: BaseModalProps) {
```
Comment (situchan) [inline src/components/Modal/index.tsx:10]: Please define default values same as the ones in defaultProps before migration.
i.e. `onModalHide()` will crash in `hideModal` function

**Review by situchan — CHANGES_REQUESTED** [review]
Mostly looks good

File: src/components/Modal/index.tsx (hunk 16 above)
Comment (situchan) [reply src/components/Modal/index.tsx:10]: Good example of regression because of `fullscreen` not defaulting to true:

Before:

Simulator Screenshot - iPhone 15

After:

Simulator Screenshot - iPhone 15

**Comment by JKobrynski** [comment]
@situchan default props brought back 😄

**Comment by situchan** [comment]
final test today

**Comment by mountiny** [comment]
@situchan let me know in slack when this is ready, thanks!

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/108292595/b53bf2eb-e7af-4eef-98cc-02370294608c



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/108292595/d02db970-2a7e-480d-b5ea-14a2f0ec78f3



</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/108292595/91336b53-a2dc-4502-b6e4-7b4eb4d921fb




</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/108292595/11469edb-ac5d-415f-ba7c-cd791ac992ca




</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/108292595/c5ed699f-f030-4687-abf8-08bf6ae4540e



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/108292595/e9003a13-6a37-4b5a-ad64-6d83c6fd01bf



</details>

**Review by situchan — APPROVED** [review]
:shipit:

File: src/components/Modal/index.tsx [hunk 17]
```diff
@@ -20,33 +20,37 @@ function Modal(props) {
...
+
+        const isFullScreenModal = type === CONST.MODAL.MODAL_TYPE.CENTERED || type === CONST.MODAL.MODAL_TYPE.CENTERED_UNSWIPEABLE || type === CONST.MODAL.MODAL_TYPE.RIGHT_DOCKED;
+
+        if (statusBarColor) {
+            setPreviousStatusBarColor(statusBarColor);
+            // If it is a full screen modal then match it with appBG, otherwise we use the backdrop color
+            setStatusBarColor(isFullScreenModal ? themeColors.appBG : StyleUtils.getThemeBackgroundColor(statusBarColor));
+        }
+
+        onModalShow?.();
```
Comment (mountiny) [inline src/components/Modal/index.tsx:37]: How does this syntax work for a method? call the method if it is not null?

**Review by mountiny — APPROVED** [review]
One commment for my understanding @JKobrynski

File: src/components/Modal/index.tsx (hunk 17 above)
Comment (situchan) [reply src/components/Modal/index.tsx:37]: Ignores when `onModalShow` is `undefined` or `null` and returns `undefined`. Otherwise same as `onModalShow()`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Comment (BartoszGrajdek) [reply src/components/Modal/index.tsx:37]: Exactly! It's very useful and you can also use it to safely access arrays i.e. lets say you have a variable `test` and it's type is `string[] | null` if you use `test?.[0]` it won't throw an error even if `test === null`, if you don't use optional chaining however this would throw a TypeError in that case

Comment (mountiny) [reply src/components/Modal/index.tsx:37]: Thanks!
